### PR TITLE
Adding a sequence number to check to identify retransmits

### PR DESCRIPTION
**Pull request by amitlimaye**
----

#### Description
*Changes proposed in this pull request.*

#### Test plan
*Outline the test plan used to test this change before merging it.*

> Fixes #.

### CODE REVIEW

File: controller/pkg/connection/connection.go [hunk 1]
```diff
@@ -201,6 +202,21 @@ func (c *TCPConnection) GetState() TCPFlowState {
...
 
+// GetInitialSequenceNumber returns the initial sequence number that was found on the syn packet
+// corresponding to this TCP Connection
+func (c *TCPConnection) GetInitialSequenceNumber() uint32 {
+	c.RLock()
+	defer c.RUnlock()
+	return c.initialSequenceNumber
+}
+
+// SetInitialSequenceNumber sets the initial sequence number from the syn packet which is creating the new connection
```
Comment (satyamsi) [inline controller/pkg/connection/connection.go]: dont need this.

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go [hunk 2]
```diff
@@ -1120,7 +1120,7 @@ func (d *Datapath) netSynRetrieveState(p *packet.Packet) (*connection.TCPConnect
 
 	context, err := d.contextFromIP(false, p.Mark, p.DestPort(), packet.IPProtocolTCP)
 	if err == nil {
-		if conn, err := d.netOrigConnectionTracker.GetReset(p.L4FlowHash(), 0); err == nil && !conn.(*connection.TCPConnection).MarkForDeletion {
+		if conn, err := d.netOrigConnectionTracker.GetReset(p.L4FlowHash(), 0); err == nil && !conn.(*connection.TCPConnection).MarkForDeletion && conn.(*connection.TCPConnection).GetInitialSequenceNumber() == p.TCPSequenceNumber() {
```
Comment (satyamsi) [inline controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: why does MarkForDeletion not need a lock to access but isn does inside. a function called GetISN ?

**Review by satyamsi — CHANGES_REQUESTED** [review]

File: controller/internal/enforcer/nfqdatapath/datapath_tcp.go (hunk 2 above)
Comment (amitlimaye) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: removed the lock from sequence number i had provided a set so added a lock. Removed the set function as well 
We are setting this only on create so no lock required. 
The markforDeleteion is modified under lock ideally this place should have a Rlock on it

Comment (amitlimaye) [reply controller/internal/enforcer/nfqdatapath/datapath_tcp.go]: Added Rlock for MarkForDeletion flag. the set happens under a global flag which is locked

**Comment by amitlimaye** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "",
    "component": "enforcerd",
    "pr-id": "1884",
    "commit-sha": "60ba4280b839cc555bc708015f97062c4eb80a21",
    "pipeline": "release-3.14.0"
  },
  {
    "project": "",
    "component": "trireme-lib",
    "pr-id": "1029",
    "commit-sha": "5c3d487b399b1128c3b8f7799a2f840f7fd868a9",
    "pipeline": "release-3.14.0"
  }
]
```

**Review by sibicramesh — APPROVED** [review]

File: controller/pkg/connection/connection.go [hunk 3]
```diff
@@ -201,6 +202,19 @@ func (c *TCPConnection) GetState() TCPFlowState {
 	return c.state
 }
 
+// GetInitialSequenceNumber returns the initial sequence number that was found on the syn packet
+// corresponding to this TCP Connection
+func (c *TCPConnection) GetInitialSequenceNumber() uint32 {
+	return c.initialSequenceNumber
```
Comment (satyamsi) [inline controller/pkg/connection/connection.go:208]: Why doesnt this need a read lock when mark for deletion does ?

**Pull request merged**